doc/ext/search.md

```diff
@@ -0,0 +1,109 @@
+# search
+
+This is a work-in-progress specification.
+
+## Description
+
+This document describes the format of the `search` extension. This enables clients to run a server-side search of messages according to specified selectors.
+
+This specification lets clients run an efficient search query on a bouncer or server who has quick access to the client message history, instead of having to download all logs and run the search locally.
+
+The server as mentioned in this document may refer to either an IRC server or an IRC bouncer.
+
+## Implementation
+
+The `search` extension uses the `soju.im/search` capability and introduces a new command, `SEARCH`, and batch type, `search`.
+
+Full support for this extension requires support for the batch, server-time and message-tags capabilities. However, limited functionality is available to clients without support for these CAPs. Servers SHOULD NOT enforce that clients support all related capabilities before using the search extension.
+
+The `soju.im/search` capability MUST be negotiated.
+
+### `SEARCH` Command
+
+The client can request a message search by sending the `SEARCH` command to the server. This command has the following general syntax:
+
+    SEARCH <attributes>
+
+If the batch capability was negotiated, the server MUST reply to a successful SEARCH command using a batch with batch type `search`. If no content exists to return, the server SHOULD return an empty batch in order to avoid the client waiting for a reply.
+
+The server then replies with a batch of batch type `search` containing messages matching all the specified attributes.
+
+### Returned message notes
```

Comment:
TODO: specify which message types we want to return (privmsg, notice, no tagmsg?)